File: reference/dotnet/Domain/Company.Product.Domain.UseCases/Types/Animal.cs
using System;

namespace Company.Product.Domain.UseCases.Types
{

    public abstract class Animal
    {
        public Guid AnimalId { get; }

        public string Sound { get; }

        protected Animal(Guid animalId, string sound)
        {
            AnimalId = animalId;
            Sound = sound;
        }

        public abstract T Visit<T>(AnimalVisitor<T> visitor);
    }
}

File: reference/springboot/domain-usecases/src/main/java/org/example/domain/usecases/types/Animal.java
package org.example.domain.usecases.types;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.UUID;

@Getter
@AllArgsConstructor
public abstract class Animal {
    private final UUID animalId;
    private final String sound;

    public abstract <T> T visit(AnimalVisitor<T> visitor);
}
